core/api/service/api_service.cpp

```diff
@@ -36,16 +57,54 @@ namespace kagome::api {
     for (const auto &listener : listeners_) {
       auto on_new_session =
           [wp = weak_from_this()](const sptr<Session> &session) mutable {
-            if (auto self = wp.lock(); not self) {
+            auto self = wp.lock();
+            if (!self) {
               return;
             }
+
+            if (SessionType::kSessionType_Ws == session->type()) {
+              auto subscribed_session =
+                  self->store_session_with_id(session->id(), session);
+              subscribed_session->set_callback([wp](SessionPtr &session,
+                                                    const auto &key,
+                                                    const auto &data,
+                                                    const auto &block) {
+                if (auto self = wp.lock()) {
+                  jsonrpc::Value::Array out_data;
+                  out_data.emplace_back(api::makeValue(key));
+                  out_data.emplace_back(api::makeValue(data));
+
+                  /// TODO(iceseer): make event notofication depending in
+                  /// blocks, to butch them in a single message
+
+                  jsonrpc::Value::Struct result;
+                  result["changes"] = std::move(out_data);
+                  result["block"] = api::makeValue(block);
+
+                  self->server_->processJsonData(
+                      result, [&](const std::string &response) {
+                        session->respond(response);
+                      });
+                }
+              });
+            }
+
             session->connectOnRequest(
                 [wp](std::string_view request,
                      std::shared_ptr<Session> session) mutable {
                   auto self = wp.lock();
-                  if (not self) {
-                    return;
-                  }
+                  if (not self) return;
+
+                  auto thread_session_auto_release = [](void *) {
+                    threaded_info.release_thread_session_id();
+                  };
+
+                  threaded_info.store_thread_session_id(session->id());
+                  // NOLINTNEXTLINE(cppcoreguidelines-pro-type-reinterpret-cast)
+                  std::unique_ptr<void, decltype(thread_session_auto_release)>
```

Comment:
Maybe gsl::finally would be more readable here.

Reply:
Because honestly although I understand this code after, it gives me creeps.

Reply:
```gsl::finally``` uses additional move object ctor and dctor call of the lambda-context. Even though the ```unique_ptr``` looks scarier, it is preferable. I guess ```gsl::finally``` creates some local wrapper over functor and then it moves it to member. I don't reversed it yet in such details.